File: app/src/main/java/com/iutcalendar/mainpage/PageEventActivity.java
package com.iutcalendar.mainpage;

import android.graphics.Color;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.widget.LinearLayout;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.FragmentTransaction;
import androidx.viewpager.widget.ViewPager;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdView;
import com.google.android.gms.ads.MobileAds;
import com.iutcalendar.calendrier.Calendrier;
import com.iutcalendar.calendrier.CurrentDate;
import com.iutcalendar.calendrier.EventCalendrier;
import com.iutcalendar.data.DataGlobal;
import com.iutcalendar.data.FileGlobal;
import com.iutcalendar.event.changement.ChangeDialog;
import com.iutcalendar.mainpage.ui.main.SectionsPagerAdapter;
import com.iutcalendar.menu.MenuItemClickActivities;
import com.iutcalendar.service.ForegroundServiceUpdate;
import com.iutcalendar.settings.SettingsApp;
import com.iutcalendar.swiping.GestureEventManager;
import com.iutcalendar.swiping.TouchGestureListener;
import com.iutcalendar.task.PersonnalCalendrier;
import com.iutcalendar.widget.WidgetCalendar;
import com.univlyon1.tools.agenda.R;
import com.univlyon1.tools.agenda.databinding.ActivityPageEventBinding;

import java.io.File;
import java.util.GregorianCalendar;

public class PageEventActivity extends AppCompatActivity {

    private static boolean updating = false;
    private ActivityPageEventBinding binding;
    private FragmentTransaction fragmentTransaction;
    private CurrentDate currDate;
    private TextView currDateLabel;
    private LinearLayout nameDayLayout;
    private LinearLayout dayOfWeek;
    private Calendrier calendrier;

    private ViewPager viewPager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Log.d("Global", "PageEventActivity start");

        if (SettingsApp.adapteTheme(this)) {//it will restart, I don't know why
            return;
        }
        SettingsApp.setLocale(getResources(), DataGlobal.getLanguage(getApplicationContext()));


        binding = ActivityPageEventBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());


        initVariable();
        PersonnalCalendrier.getInstance(getApplicationContext());

        initActionBar();

        //init Calendrier
        initCalendar();


        setCurrDate(getDateToLaunchAtFirst());

        initPageViewEvent();


        initGestureSwipeWeek();


        initAllOnClickEvents();


        //affiche la dialog si ouverte depuis notification (changements)
        String changements = getIntent().getStringExtra("changes");
        if (changements != null) {
            Log.d("Extra", changements);
            showChangedEvent(changements);
        } else {
            Log.d("Extra", "no changes");
        }

        //démarre le service d'arrière-plan avec interval
        ForegroundServiceUpdate.start(getApplicationContext());

        //update
        update(null);

        initAds();
        /*####Testing feature#####*/

        Log.d("Global", "PageEventActivity end");


    }


    private void initAds() {
        MobileAds.initialize(this, initializationStatus -> {
        });

        AdView mAdView = findViewById(R.id.adView);
        AdRequest adRequest = new AdRequest.Builder().build();
        mAdView.loadAd(adRequest);
    }

    /**
     * init all on click events
     */
    private void initAllOnClickEvents() {
        //back to current date
        currDateLabel.setOnClickListener(view -> setCurrDate(new CurrentDate()));


        //Click on day
        setOnClickDay(binding.dayLundi, 0);
        setOnClickDay(binding.dayMardi, 1);
        setOnClickDay(binding.dayMercredi, 2);
        setOnClickDay(binding.dayJeudi, 3);
        setOnClickDay(binding.dayVendredi, 4);
        setOnClickDay(binding.daySamedi, 5);
        setOnClickDay(binding.dayDimanche, 6);


        setOnClickDay(binding.lundiNum, 0);
        setOnClickDay(binding.mardiNum, 1);
        setOnClickDay(binding.mercrediNum, 2);
        setOnClickDay(binding.jeudiNum, 3);
        setOnClickDay(binding.vendrediNum, 4);
        setOnClickDay(binding.samediNum, 5);
        setOnClickDay(binding.dimancheNum, 6);
    }

    /**
     * init gesture to switch between weeks
     */
    private void initGestureSwipeWeek() {
        int childCount = nameDayLayout.getChildCount();
        for (int i = 0; i < childCount; i++) {
            nameDayLayout.getChildAt(i).setOnTouchListener(new TouchGestureListener(getApplicationContext(), new PageEventActivity.GestureWeekListener()));
        }
        nameDayLayout.setOnTouchListener(new TouchGestureListener(getApplicationContext(), new PageEventActivity.GestureWeekListener()));
        dayOfWeek.setOnTouchListener(new TouchGestureListener(getApplicationContext(), new PageEventActivity.GestureWeekListener()));
    }

    /**
     * init viewPager and sectionsPagerAdapter to see events
     */
    private void initPageViewEvent() {
        Log.d("Event", "cration Section page adapter------------------");

        viewPager = binding.viewPager;
        SectionsPagerAdapter sectionsPagerAdapter = new SectionsPagerAdapter(this, getSupportFragmentManager(), getCalendrier());
        viewPager.setAdapter(sectionsPagerAdapter);
        viewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
            }

            @Override
            public void onPageSelected(int position) {
                Log.d("Page", "new page : " + position);
                setCurrDate(new CurrentDate(getCalendrier().getFirstDay()).addDay(position));
            }

            @Override
            public void onPageScrollStateChanged(int state) {
            }
        });
        setPositionPageToCurrDate();
    }

    private void updatePageViewEvent() {
        Log.d("Event", "update Section page adapter------------------");
        SectionsPagerAdapter sectionsPagerAdapter = new SectionsPagerAdapter(this, getSupportFragmentManager(), getCalendrier());
        viewPager.removeAllViews();
        viewPager.setAdapter(sectionsPagerAdapter);
        setPositionPageToCurrDate();
    }

    /**
     * @return the date the activity has to be launch
     */
    private CurrentDate getDateToLaunchAtFirst() {
        CurrentDate dateToLaunch = new CurrentDate();
        Log.d("Widget", "main : " + getIntent().getBooleanExtra("launch_next_event", false));
        if (getIntent().getBooleanExtra("launch_next_event", false)) {
            dateToLaunch.add(GregorianCalendar.MINUTE, WidgetCalendar.DELAY_AFTER_EVENT_PASSED);//pcq l'event s'affiche toujours au bout de 30min
            EventCalendrier[] es = getCalendrier().getNext2EventAfter(dateToLaunch);
            if (es[0] != null) {
                dateToLaunch = new CurrentDate(es[0].getDate());
            }
        } else {
            Log.d("Widget", "default date");
        }
        return dateToLaunch;
    }

    public ActivityPageEventBinding getBinding() {
        return binding;
    }

    private void initCalendar() {
        File fileCal = FileGlobal.getFileDownload(getApplicationContext());
        String str = FileGlobal.readFile(fileCal);
        setCalendrier(new Calendrier(str));
    }

    private void initVariable() {
        this.currDate = new CurrentDate();

        this.fragmentTransaction = getSupportFragmentManager().beginTransaction();

        this.nameDayLayout = findViewById(R.id.nameDayLayout);
        this.dayOfWeek = findViewById(R.id.dayOfWeek);
    }

    private void initActionBar() {
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayOptions(ActionBar.DISPLAY_SHOW_CUSTOM);
            actionBar.setCustomView(R.layout.action_bar_main_page);
            currDateLabel = actionBar.getCustomView().findViewById(R.id.title);
        }
    }



    /*########################################################################
                                     UPDATE
    ########################################################################*/

    /**
     * update the calendar file
     *
     * @param onFinishListener est toujours appelé à la fin de la méthode
     */
    public void update(OnFinishListener onFinishListener) {
        if (!updating) {
            Log.d("Update", "start");
            updating = true;
            new Thread(() -> {
                FileGlobal.updateAndGetChange(this, calendrier, ((context, intent) -> startActivity(intent)));

                runOnUiThread(() -> {
                    Log.d("Debug", "appeler d'epuis update");

                    updatePageViewEvent();
//                    setCurrDate(getCurrDate());
                });

                if (onFinishListener != null) onFinishListener.finished();
                updating = false;
                Log.d("Update", "end");
            }).start();
        }


    }

    private void showChangedEvent(String changes) {
        //TODO faire plus beau
        //FIXME pas afficher quand charger agenda
//        AlertDialog.Builder alertDialog = new AlertDialog.Builder(PageEventActivity.this);
//        alertDialog.setTitle(getString(R.string.event));
//
////        alertDialog.setMessage(changes);
//        alertDialog.setView(getLayoutInflater().inflate(R.layout.fragment_history, null));
//
//        alertDialog.setPositiveButton(getString(R.string.ok), (dialog, which) -> dialog.dismiss());
//
//        alertDialog.show();

        new ChangeDialog(this).show();

    }

    /*########################################################################
                               GETTERS & SETTERS
    ########################################################################*/
    public CurrentDate getCurrDate() {
        return this.currDate;
    }

    public void setCurrDate(CurrentDate newDate) {
//        if (newDate.sameDay(currDate)) { //some bugs
//            currDateLabel.setText(currDate.getRelativeDayName(getBaseContext()));
//            return;
//        }
        Log.d("Date", currDate + " set curr date to " + newDate);

        if (getCalendrier() != null && getCalendrier().getFirstDay() != null && getCalendrier().getLastDay() != null) {
            if (newDate.compareTo(getCalendrier().getFirstDay()) < 0) {
                newDate = new CurrentDate(getCalendrier().getFirstDay());
            } else if (newDate.compareTo(getCalendrier().getLastDay()) > 0) {
                newDate = new CurrentDate(getCalendrier().getLastDay());
            }
        }

        this.currDate.set(newDate);

        currDateLabel.setText(currDate.getRelativeDayName(getBaseContext()));
        Log.d("ActionBar", currDate.getRelativeDayName(getBaseContext()));

        setPositionPageToCurrDate();
        setDaysOfWeek();
    }

    private void setPositionPageToCurrDate() {
        if (getCalendrier() != null && getCalendrier().getFirstDay() != null && viewPager != null) {
            viewPager.setCurrentItem(getPosOfDate(this.currDate));
        }
    }

    private int getPosOfDate(CurrentDate date) {
        int pos = getCalendrier().getFirstDay().getNbrDayTo(date);
        Log.d("Event", "position = " + pos + " for date " + date);
        Log.d("Position", "offset : " + pos);
        return pos;
    }

    public Calendrier getCalendrier() {
        return calendrier;
    }

    public void setCalendrier(Calendrier calendrier) {
        this.calendrier = calendrier;
    }

    public void setDaysOfWeek() {
        setNumOfMonthAndSelected(binding.lundiNum, 0);
        setNumOfMonthAndSelected(binding.mardiNum, 1);
        setNumOfMonthAndSelected(binding.mercrediNum, 2);
        setNumOfMonthAndSelected(binding.jeudiNum, 3);
        setNumOfMonthAndSelected(binding.vendrediNum, 4);
        setNumOfMonthAndSelected(binding.samediNum, 5);
        setNumOfMonthAndSelected(binding.dimancheNum, 6);
    }

    public void setAnimationGoRight() {
        fragmentTransaction.setCustomAnimations(R.anim.enter_from_right, R.anim.exit_to_left);
    }

    public void setAnimationGoLeft() {
        fragmentTransaction.setCustomAnimations(R.anim.enter_from_left, R.anim.exit_to_right);

    }

    public void setAnimationDirection(int d) {
        if (d > 0) {
            setAnimationGoRight();
        } else {
            setAnimationGoLeft();
        }
    }

    private void setOnClickDay(TextView dayClicked, int day) {
        if (dayClicked != null)
            dayClicked.setOnTouchListener(new TouchGestureListener(getApplicationContext(), new PageEventActivity.GestureWeekListener() {
                @Override
                protected void onClick() {
                    setCurrDate(currDate.getDateOfDayOfWeek(day));
                    super.onClick();
                }
            }));

    }

    /**
     * met les numéros des jours et leur couleur de fond adapté
     *
     * @param numDay le label du jour
     * @param day    son numéro de la semaine
     */
    public void setNumOfMonthAndSelected(TextView numDay, int day) {
        numDay.setText(String.valueOf(getCurrDate().getDateOfDayOfWeek(day).getDay()));


        if (getCurrDate().getDay() == getCurrDate().getDateOfDayOfWeek(day).getDay()) {
            numDay.setBackgroundColor(Color.parseColor("#FFC41442"));
        } else if (getCurrDate().getDateOfDayOfWeek(day).sameDay(new CurrentDate())) {
            numDay.setBackgroundColor(Color.parseColor("#88C41442"));
        } else {
            numDay.setBackgroundColor(Color.argb(0f, 1.0f, 1.0f, 1.0f));
        }
    }

    /*########################################################################
                                    NAVIGATION WEEKS
     ########################################################################*/

    public void switchToPrevWeek() {
        int sens = -1;

        setAnimationDirection(sens);
        setCurrDate(getCurrDate().prevWeek());
    }

    public void switchToNextWeek() {
        int sens = -1;

        setAnimationDirection(sens);
        setCurrDate(getCurrDate().nextWeek());
    }

    /*########################################################################
                                    ACTION BAR
     ########################################################################*/
    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        return new MenuItemClickActivities(this).onMenuItemClick(item);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.menu_activities, menu);
        return super.onCreateOptionsMenu(menu);
    }


    /*########################################################################
                                    GESTION EVENT
     ########################################################################*/

    private class GestureWeekListener extends GestureEventManager {

        @Override
        public void onSwipeRight() {
            super.onSwipeRight();
            switchToPrevWeek();
        }

        @Override
        public void onSwipeLeft() {
            super.onSwipeLeft();
            switchToNextWeek();
        }
    }
}

File: app/src/main/java/com/iutcalendar/mainpage/PageEventActivity.kt
package com.iutcalendar.mainpage

import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.widget.TextView
import androidx.appcompat.app.ActionBar
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.FragmentTransaction
import androidx.lifecycle.lifecycleScope
import androidx.viewpager2.widget.ViewPager2
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.MobileAds
import com.iutcalendar.calendrier.Calendrier
import com.iutcalendar.calendrier.CurrentDate
import com.iutcalendar.data.CachedData
import com.iutcalendar.data.DataGlobal
import com.iutcalendar.data.FileGlobal
import com.iutcalendar.dialog.DialogMessage
import com.iutcalendar.event.changement.ChangeDialog
import com.iutcalendar.mainpage.ui.main.SectionsPagerAdapter
import com.iutcalendar.menu.MenuItemClickActivities
import com.iutcalendar.service.WorkUpdate
import com.iutcalendar.settings.SettingsApp
import com.iutcalendar.swiping.GestureEventManager
import com.iutcalendar.swiping.TouchGestureListener
import com.iutcalendar.task.PersonalCalendrier
import com.iutcalendar.widget.WidgetCalendar
import com.univlyon1.tools.agenda.BuildConfig
import com.univlyon1.tools.agenda.R
import com.univlyon1.tools.agenda.databinding.ActivityPageEventBinding
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.util.*


class PageEventActivity : AppCompatActivity() {
    lateinit var binding: ActivityPageEventBinding
        private set
    private var fragmentTransaction: FragmentTransaction? = null
    private var currDate: CurrentDate = CurrentDate()
    private var currDateLabel: TextView? = null
    var calendrier: Calendrier
        get() = CachedData.calendrier
        set(value) {
            CachedData.calendrier = value
        }
    private val viewPager: ViewPager2
        get() = binding.viewPager

    var savedInstanceCalendrier: Calendrier? = null


    override fun onResume() {
        super.onResume()
        Log.d("PageEventActivity", "onResume")
        if (hasToReload) {
            currDate = CurrentDate()
            initPageViewEvent()
            hasToReload = false
        } else savedInstanceCalendrier?.let {
            if (it != CachedData.calendrier) {
                initPageViewEvent()
            }
            Log.d("PageEventActivity", "new has = ${CachedData.calendrier.events.size}\t prev has ${it.events.size}")
        }
    }

    override fun onPause() {
        super.onPause()
        savedInstanceCalendrier = calendrier.clone()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        Log.d("PageEventActivity", "PageEventActivity start")
        binding = ActivityPageEventBinding.inflate(layoutInflater)
        setContentView(binding.root)
        initVariable()
        PersonalCalendrier.getInstance(applicationContext)
        initActionBar()

        //init Calendrier
        initCalendar()
        setCurrDate(dateToLaunchAtFirst)
        initPageViewEvent()
        initGestureSwipeWeek()
        initAllOnClickEvents()


        //affiche la dialog si changements depuis la dernière fois (notif ou non)
        var nombreChange: Int
        if (DataGlobal.getSavedInt(this, DataGlobal.NOMBRE_CHANGE_TO_DISPLAY)
                .also { nombreChange = it } > 0
        ) {
            showChangedEvent(nombreChange)
            DataGlobal.save(this, DataGlobal.NOMBRE_CHANGE_TO_DISPLAY, 0)
        } else {
            Log.d("Extra", "no changes")
        }

        //démarre le service en arrière-plan avec interval
        WorkUpdate.startBackgroundWork(this)

        //update
        update {}
        initAds()

        //if update show news
        if (DataGlobal.getSavedString(applicationContext, DataGlobal.LAST_UPDATE_TAG) != BuildConfig.VERSION_NAME) {
            DialogMessage.showInfo(this, baseContext.resources.getString(R.string.What_new), baseContext.resources.getString(R.string.news_msg)) {
                DataGlobal.save(applicationContext, DataGlobal.LAST_UPDATE_TAG, BuildConfig.VERSION_NAME)
            }
        }


        /*####Testing feature#####*/
        Log.d("PageEventActivity", "PageEventActivity end")

    }

    private fun initAds() {
        MobileAds.initialize(this) { }
        binding.adView.loadAd(AdRequest.Builder().build())
    }

    /**
     * init all on click events
     */
    private fun initAllOnClickEvents() {
        //back to current date
        currDateLabel?.setOnClickListener { setCurrDate(CurrentDate()) }

        //Click on day
        binding.let {
            setOnClickDay(it.dayLundi, 0)
            setOnClickDay(it.dayMardi, 1)
            setOnClickDay(it.dayMercredi, 2)
            setOnClickDay(it.dayJeudi, 3)
            setOnClickDay(it.dayVendredi, 4)
            setOnClickDay(it.daySamedi, 5)
            setOnClickDay(it.dayDimanche, 6)

            setOnClickDay(it.lundiNum, 0)
            setOnClickDay(it.mardiNum, 1)
            setOnClickDay(it.mercrediNum, 2)
            setOnClickDay(it.jeudiNum, 3)
            setOnClickDay(it.vendrediNum, 4)
            setOnClickDay(it.samediNum, 5)
            setOnClickDay(it.dimancheNum, 6)
        }
    }

    /**
     * init gesture to switch between weeks
     */
    private fun initGestureSwipeWeek() {
        binding.nameDayLayout.apply {
            for (i in 0 until childCount) {
                getChildAt(i)
                    .setOnTouchListener(TouchGestureListener(applicationContext, GestureWeekListener()))
            }
            setOnTouchListener(
                TouchGestureListener(
                    applicationContext,
                    GestureWeekListener()
                )
            )
        }
        binding.dayOfWeek.apply {
            setOnTouchListener(
                TouchGestureListener(
                    applicationContext,
                    GestureWeekListener()
                )
            )
        }
    }

    /**
     * init viewPager and sectionsPagerAdapter to see events
     */
    fun initPageViewEvent() {
        Log.d("Event", "creation Section page adapter")
        val sectionsPagerAdapter = SectionsPagerAdapter(this)
        viewPager.adapter = sectionsPagerAdapter
        viewPager.registerOnPageChangeCallback(object : ViewPager2.OnPageChangeCallback() {
            override fun onPageSelected(position: Int) {
                Log.d("Page", "new page : $position")
                calendrier.firstDay?.let {
                    setCurrDate(CurrentDate(it).addDay(position))
                }

            }
        })

        setCurrDate(currDate)
        setPositionPageToCurrDate()
    }


    private fun updatePageViewEvent() {
        Log.d("Event", "update Section page adapter")

        viewPager.adapter?.notifyDataSetChanged()

        setPositionPageToCurrDate()
    }

    /**
     * @return the date the activity has to be launch
     */
    private val dateToLaunchAtFirst: CurrentDate
        get() {
            var dateToLaunch = CurrentDate()
            Log.d("Widget", "main : " + intent.getBooleanExtra("launch_next_event", false))
            if (intent.getBooleanExtra("launch_next_event", false)) {
                dateToLaunch.add(
                    GregorianCalendar.MINUTE,
                    WidgetCalendar.DELAY_AFTER_EVENT_PASSED
                ) //pcq event est toujours affiché toujours au bout de 30min
                val es = calendrier.getNext2EventAfter(dateToLaunch)
                es[0]?.date?.let { date ->
                    dateToLaunch = CurrentDate(date)
                }
            } else {
                Log.d("Widget", "default date")
            }
            return dateToLaunch
        }

    private fun initCalendar() {
        calendrier = Calendrier(FileGlobal.readFile(FileGlobal.getFileCalendar(applicationContext)))
    }

    private fun initVariable() {
        currDate = CurrentDate()
        fragmentTransaction = supportFragmentManager.beginTransaction()
    }

    private fun initActionBar() {
        val actionBar = supportActionBar
        if (actionBar != null) {
            actionBar.displayOptions = ActionBar.DISPLAY_SHOW_CUSTOM
            actionBar.setCustomView(R.layout.action_bar_main_page)
            currDateLabel = actionBar.customView.findViewById(R.id.title)
        }
    }
    /*########################################################################
                                     UPDATE
    ########################################################################*/

    private val poolActionEndUpdate = ArrayDeque<() -> Unit>()

    /**
     * update the calendar file
     *
     * @param onFinishListener toujours appelé quand la maj se fini
     */
    fun update(onFinishListener: (() -> Unit)) {
        poolActionEndUpdate.add(onFinishListener)
        if (!updating) {
            Log.d("Update", "start")
            updating = true

            lifecycleScope.launch(Dispatchers.IO) {
                FileGlobal.updateAndGetChange(
                    this@PageEventActivity,
                    calendrier
                ) { _: Context?, intent: Intent? -> startActivity(intent) }

                withContext(Dispatchers.Main) {
                    updatePageViewEvent()
                }
                while (poolActionEndUpdate.isNotEmpty()) {
                    poolActionEndUpdate.pop()()
                }
                updating = false
                Log.d("Update", "end")
            }
        }
    }


    private fun showChangedEvent(nombreChangements: Int) {
        ChangeDialog(this, nombreChangements).show()
    }

    /*########################################################################
                               GETTERS & SETTERS
    ########################################################################*/

    fun setCurrDate(newCurrentDate: CurrentDate?) {
        if (newCurrentDate != null) {
            var newDate: CurrentDate = newCurrentDate
            Log.d("Date", "$currDate set curr date to $newDate")
            calendrier.firstDay?.let { firstDay ->
                calendrier.lastDay?.let { lastDay ->
                    if (newDate < firstDay) {
                        newDate = CurrentDate(firstDay)
                    } else if (newDate > lastDay) {
                        newDate = CurrentDate(lastDay)
                    }
                }
            }
            currDate.set(newDate)
            currDateLabel!!.text = currDate.getRelativeDayName(baseContext)
            Log.d("ActionBar", currDate.getRelativeDayName(baseContext))
            setPositionPageToCurrDate()
            setDaysOfWeek()
        }

    }

    private fun setPositionPageToCurrDate() {
        if (calendrier.firstDay != null) {
            viewPager.currentItem = getPosOfDate(currDate)
        }
    }

    private fun getPosOfDate(date: CurrentDate?): Int {
        if (calendrier.firstDay != null) {
            val pos = calendrier.firstDay!!.getNbrDayTo(date)
            Log.d("Event", "position = $pos for date $date")
            Log.d("Position", "offset : $pos")
            return pos
        }
        return 0
    }

    private fun setDaysOfWeek() {
        setNumOfMonthAndSelected(binding.lundiNum, 0)
        setNumOfMonthAndSelected(binding.mardiNum, 1)
        setNumOfMonthAndSelected(binding.mercrediNum, 2)
        setNumOfMonthAndSelected(binding.jeudiNum, 3)
        setNumOfMonthAndSelected(binding.vendrediNum, 4)
        setNumOfMonthAndSelected(binding.samediNum, 5)
        setNumOfMonthAndSelected(binding.dimancheNum, 6)
    }

    private fun setOnClickDay(dayClicked: TextView?, day: Int) {
        dayClicked?.apply {
            setOnTouchListener(
                TouchGestureListener(
                    applicationContext,
                    object : GestureWeekListener() {
                        override fun onClick() {
                            setCurrDate(currDate.getDateOfDayOfWeek(day))
                            super.onClick()
                        }
                    })
            )
        }
    }

    /**
     * Met les numéros des jours et leur couleur de fond adapté
     *
     * @param numDay le label du jour
     * @param day    son numéro de la semaine
     */
    private fun setNumOfMonthAndSelected(numDay: TextView, day: Int) {
        numDay.text = currDate.getDateOfDayOfWeek(day).day.toString()
        if (currDate.day == currDate.getDateOfDayOfWeek(day).day) {
            numDay.setBackgroundColor(Color.parseColor("#FFC41442"))
        } else if (currDate.getDateOfDayOfWeek(day).sameDay(CurrentDate())) {
            numDay.setBackgroundColor(Color.parseColor("#88C41442"))
        } else {
            numDay.setBackgroundColor(Color.argb(0f, 1.0f, 1.0f, 1.0f))
        }
    }

    /*########################################################################
                                    NAVIGATION WEEKS
     ########################################################################*/
    fun switchToPrevWeek() {
        setCurrDate(currDate.prevWeek())
    }

    fun switchToNextWeek() {
        setCurrDate(currDate.nextWeek())
    }

    /*########################################################################
                                    ACTION BAR
     ########################################################################*/
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return MenuItemClickActivities(this).onMenuItemClick(item)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.menu_activities, menu)
        return super.onCreateOptionsMenu(menu)
    }

    /*########################################################################
                                    GESTION EVENT
     ########################################################################*/
    private open inner class GestureWeekListener : GestureEventManager() {
        override fun onSwipeRight() {
            super.onSwipeRight()
            switchToPrevWeek()
        }

        override fun onSwipeLeft() {
            super.onSwipeLeft()
            switchToNextWeek()
        }
    }

    companion object {
        private var updating = false
        var hasToReload = false
    }
}